File: gofer/src/storage/task_runs.rs
use std::ops::{Deref, Not};

use crate::storage::{SqliteErrors, StorageError, MAX_ROW_LIMIT};
use futures::TryFutureExt;
use gofer_models::task_run::{State, Status, StatusReason, TaskRun};
use gofer_models::Variable;
use sqlx::{sqlite::SqliteRow, QueryBuilder, Row, Sqlite, SqliteConnection};
use std::str::FromStr;

#[derive(Debug, Default)]
pub struct UpdatableFields {
    pub started: Option<u64>,
    pub ended: Option<u64>,
    pub exit_code: Option<u8>,
    pub failure: Option<StatusReason>,
    pub logs_expired: Option<bool>,
    pub logs_removed: Option<bool>,
    pub state: Option<State>,
    pub status: Option<Status>,
    pub scheduler_id: Option<String>,
    pub variables: Option<Vec<Variable>>,
}

/// Return all task_run for a given namespace/pipeline/run; limited to 200 rows per response.
pub async fn list(
    conn: &mut SqliteConnection,
    offset: u64,
    limit: u64,
    namespace_id: &str,
    pipeline_id: &str,
    run_id: u64,
) -> Result<Vec<TaskRun>, StorageError> {
    let mut limit = limit;

    if limit == 0 || limit > MAX_ROW_LIMIT {
        limit = MAX_ROW_LIMIT;
    }

    // First we need to get the general task_run information.
    let task_runs = sqlx::query(
            r#"
SELECT namespace, pipeline, run, id, task, created, started, ended, exit_code, failure,
logs_expired, logs_removed, state, status, scheduler_id, variables
FROM task_runs
WHERE namespace = ? AND pipeline = ? AND run = ?
LIMIT ?
OFFSET ?;"#,
        )
        .bind(namespace_id)
        .bind(pipeline_id)
        .bind(run_id as i64)
        .bind(limit as i64)
        .bind(offset as i64)
        .map(|row: SqliteRow| TaskRun {
            namespace: row.get("namespace"),
            pipeline: row.get("pipeline"),
            run: row.get::<i64, _>("run") as u64,
            id: row.get("id"),
            task: {
                let task_json = row.get::<String, _>("task");
                serde_json::from_str(&task_json).unwrap()
            },
            created: row.get::<i64, _>("created") as u64,
            started: row.get::<i64, _>("started") as u64,
            ended: row.get::<i64, _>("ended") as u64,
            exit_code: row.get("exit_code"),
            status_reason: {
                let failure = row.get::<String, _>("failure");
                failure.is_empty().not().then(|| serde_json::from_str(&failure).unwrap())
            },
            logs_expired: {
                let logs_expired = match row.get::<i64, _>("logs_expired") {
                    0 => false,
                    1 => true,
                    _ => panic!("could not parse value into task_run logs_expired; only allowed values are 0 and 1")
                };
                logs_expired
            },
            logs_removed: {
                let logs_removed = match row.get::<i64, _>("logs_expired") {
                    0 => false,
                    1 => true,
                    _ => panic!("could not parse value into task_run logs_expired; only allowed values are 0 and 1")
                };
                logs_removed
            },
            state: State::from_str(row.get("state"))
                .map_err(|_| StorageError::Parse {
                    value: row.get("state"),
                    column: "state".to_string(),
                    err: "could not parse value into task_run state enum".to_string(),
                })
                .unwrap(),
            status: Status::from_str(row.get("status"))
                .map_err(|_| StorageError::Parse {
                    value: row.get("status"),
                    column: "status".to_string(),
                    err: "could not parse value into task_run status enum".to_string(),
                })
                .unwrap(),
            scheduler_id: row.get("scheduler_id"),
            variables: {
                let variables_json = row.get::<String, _>("variables");
                serde_json::from_str(&variables_json).unwrap()
            },
        })
        .fetch_all(conn)
        .map_err(|e| StorageError::Unknown(e.to_string()))
        .await?;

    Ok(task_runs)
}

/// Insert a new task_run.
pub async fn insert(conn: &mut SqliteConnection, task_run: &TaskRun) -> Result<(), StorageError> {
    sqlx::query(
        r#"
INSERT INTO task_runs (namespace, pipeline, run, id, task, created, started, ended,
    exit_code, failure, logs_expired, logs_removed, state, status, scheduler_id, variables)
VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?);"#,
    )
    .bind(&task_run.namespace)
    .bind(&task_run.pipeline)
    .bind(task_run.run as i64)
    .bind(&task_run.id)
    .bind(serde_json::to_string(&task_run.task).unwrap())
    .bind(task_run.created as i64)
    .bind(task_run.started as i64)
    .bind(task_run.ended as i64)
    .bind(task_run.exit_code)
    .bind(
        task_run
            .status_reason
            .is_none()
            .not()
            .then(|| serde_json::to_string(&task_run.status_reason).unwrap()),
    )
    .bind({
        let task_run_bool: i32 = match task_run.logs_expired {
            false => 0,
            true => 1,
        };

        task_run_bool
    })
    .bind({
        let task_run_bool: i32 = match task_run.logs_removed {
            false => 0,
            true => 1,
        };

        task_run_bool
    })
    .bind(task_run.state.to_string())
    .bind(task_run.status.to_string())
    .bind(&task_run.scheduler_id)
    .bind(serde_json::to_string(&task_run.variables).unwrap())
    .execute(conn)
    .map_ok(|_| ())
    .map_err(|e| match e {
        sqlx::Error::Database(database_err) => {
            if let Some(err_code) = database_err.code() {
                if err_code.deref() == SqliteErrors::Constraint.value() {
                    return StorageError::Exists;
                }
            }
            return StorageError::Unknown(database_err.message().to_string());
        }
        _ => StorageError::Unknown("".to_string()),
    })
    .await
}

/// Get details on a specific task_run.
pub async fn get(
    conn: &mut SqliteConnection,
    namespace_id: &str,
    pipeline_id: &str,
    run_id: u64,
    id: &str,
) -> Result<TaskRun, StorageError> {
    let task_run = sqlx::query(
            r#"
SELECT namespace, pipeline, run, id, task, created, started, ended, exit_code, failure,
logs_expired, logs_removed, state, status, scheduler_id, variables
FROM task_runs
WHERE namespace = ? AND pipeline = ? AND run = ? AND id = ?;"#,
        )
        .bind(namespace_id)
        .bind(pipeline_id)
        .bind(run_id as i64)
        .bind(id)
        .map(|row: SqliteRow| TaskRun {
            namespace: row.get("namespace"),
            pipeline: row.get("pipeline"),
            run: row.get::<i64, _>("run") as u64,
            id: row.get("id"),
            task: {
                let task_json = row.get::<String, _>("task");
                serde_json::from_str(&task_json).unwrap()
            },
            created: row.get::<i64, _>("created") as u64,
            started: row.get::<i64, _>("started") as u64,
            ended: row.get::<i64, _>("ended") as u64,
            exit_code: row.get("exit_code"),
            status_reason: {
                let failure = row.get::<String, _>("failure");
                failure.is_empty().not().then(||serde_json::from_str(&failure).unwrap())
            },
            logs_expired: {
                let logs_expired = match row.get::<i64, _>("logs_expired") {
                    0 => false,
                    1 => true,
                    _ => panic!("could not parse value into task_run logs_expired; only allowed values are 0 and 1")
                };
                logs_expired
            },
            logs_removed: {
                let logs_removed = match row.get::<i64, _>("logs_expired") {
                    0 => false,
                    1 => true,
                    _ => panic!("could not parse value into task_run logs_expired; only allowed values are 0 and 1")
                };
                logs_removed
            },
            state: State::from_str(row.get("state"))
                .map_err(|_| StorageError::Parse {
                    value: row.get("state"),
                    column: "state".to_string(),
                    err: "could not parse value into task_run state enum".to_string(),
                })
                .unwrap(),
            status: Status::from_str(row.get("status"))
                .map_err(|_| StorageError::Parse {
                    value: row.get("status"),
                    column: "status".to_string(),
                    err: "could not parse value into task_run status enum".to_string(),
                })
                .unwrap(),
            scheduler_id: row.get("scheduler_id"),
            variables: {
                let variables_json = row.get::<String, _>("variables");
                serde_json::from_str(&variables_json).unwrap()
            },
        })
        .fetch_one(conn)
        .map_err(|e| match e {
            sqlx::Error::RowNotFound => StorageError::NotFound,
            _ => StorageError::Unknown(e.to_string()),
        })
        .await?;

    Ok(task_run)
}

/// Update a specific task_run.
pub async fn update(
    conn: &mut SqliteConnection,
    task_run: &TaskRun,
    fields: UpdatableFields,
) -> Result<(), StorageError> {
    let mut update_query: QueryBuilder<Sqlite> = QueryBuilder::new(r#"UPDATE task_runs SET "#);

    let mut updated_fields_total = 0;

    if let Some(started) = fields.started {
        update_query.push("started = ");
        update_query.push_bind(started as i64);
        updated_fields_total += 1;
    }

    if let Some(ended) = fields.ended {
        if updated_fields_total > 0 {
            update_query.push(", ");
        }
        update_query.push("ended = ");
        update_query.push_bind(ended as i64);
        updated_fields_total += 1;
    }

    if let Some(exit_code) = fields.exit_code {
        if updated_fields_total > 0 {
            update_query.push(", ");
        }
        update_query.push("exit_code = ");
        update_query.push_bind(exit_code as i64);
        updated_fields_total += 1;
    }

    if let Some(failure) = fields.failure {
        if updated_fields_total > 0 {
            update_query.push(", ");
        }
        update_query.push("failure = ");
        update_query.push_bind(serde_json::to_string(&failure).unwrap());
        updated_fields_total += 1;
    }

    if let Some(logs_expired) = fields.logs_expired {
        if updated_fields_total > 0 {
            update_query.push(", ");
        }
        update_query.push("logs_expired = ");
        update_query.push_bind::<i64>({
            match logs_expired {
                false => 0,
                true => 1,
            }
        });
        updated_fields_total += 1;
    }

    if let Some(logs_removed) = fields.logs_removed {
        if updated_fields_total > 0 {
            update_query.push(", ");
        }
        update_query.push("logs_removed = ");
        update_query.push_bind::<i64>({
            match logs_removed {
                false => 0,
                true => 1,
            }
        });
        updated_fields_total += 1;
    }

    if let Some(state) = fields.state {
        if updated_fields_total > 0 {
            update_query.push(", ");
        }
        update_query.push("state = ");
        update_query.push_bind(state.to_string());
        updated_fields_total += 1;
    }

    if let Some(status) = fields.status {
        if updated_fields_total > 0 {
            update_query.push(", ");
        }
        update_query.push("status = ");
        update_query.push_bind(status.to_string());
        updated_fields_total += 1;
    }

    if let Some(scheduler_id) = fields.scheduler_id {
        if updated_fields_total > 0 {
            update_query.push(", ");
        }
        update_query.push("scheduler_id = ");
        update_query.push_bind(scheduler_id);
        updated_fields_total += 1;
    }

    if let Some(variables) = fields.variables {
        if updated_fields_total > 0 {
            update_query.push(", ");
        }
        update_query.push("variables = ");
        update_query.push_bind(serde_json::to_string(&variables).unwrap());
    }

    update_query.push(" WHERE namespace = ");
    update_query.push_bind(&task_run.namespace);

    update_query.push(" AND pipeline = ");
    update_query.push_bind(&task_run.pipeline);

    update_query.push(" AND run = ");
    update_query.push_bind(task_run.run as i64);

    update_query.push(" AND id = ");
    update_query.push_bind(&task_run.id);
    update_query.push(";");

    let update_query = update_query.build();

    update_query
        .execute(conn)
        .map_ok(|_| ())
        .map_err(|e| match e {
            sqlx::Error::RowNotFound => StorageError::NotFound,
            _ => StorageError::Unknown(e.to_string()),
        })
        .await?;

    Ok(())
}

pub async fn delete(
    conn: &mut SqliteConnection,
    namespace_id: &str,
    pipeline_id: &str,
    run_id: u64,
    id: &str,
) -> Result<(), StorageError> {
    sqlx::query(
        r#"
DELETE FROM task_runs
WHERE namespace = ? AND pipeline = ? AND run = ? AND id = ?;"#,
    )
    .bind(namespace_id)
    .bind(pipeline_id)
    .bind(run_id as i64)
    .bind(id)
    .execute(conn)
    .map_ok(|_| ())
    .map_err(|e| match e {
        sqlx::Error::RowNotFound => StorageError::NotFound,
        _ => StorageError::Unknown(e.to_string()),
    })
    .await
}

File: internal/storage/taskRuns.go
package storage

import (
	"database/sql"
	"encoding/json"
	"errors"
	"fmt"
	"strings"

	qb "github.com/Masterminds/squirrel"
	"github.com/clintjedwards/gofer/models"
)

type UpdatableTaskRunFields struct {
	Started      *int64
	Ended        *int64
	ExitCode     *int64
	State        *models.TaskRunState
	Status       *models.TaskRunStatus
	StatusReason *models.TaskRunStatusReason
	LogsExpired  *bool
	LogsRemoved  *bool
	Variables    *[]models.Variable
}

func (db *DB) ListTaskRuns(offset, limit int, namespace, pipeline string, run int64) ([]models.TaskRun, error) {
	if limit == 0 || limit > db.maxResultsLimit {
		limit = db.maxResultsLimit
	}

	rows, err := qb.Select("namespace", "pipeline", "run", "id", "task", "created", "started", "ended", "exit_code",
		"state", "status", "status_reason", "logs_expired", "logs_removed", "variables").
		From("task_runs").
		Where(qb.Eq{"namespace": namespace, "pipeline": pipeline, "run": run}).
		Limit(uint64(limit)).
		Offset(uint64(offset)).RunWith(db).Query()
	if err != nil {
		return nil, fmt.Errorf("database error occurred: %v; %w", err, ErrInternal)
	}
	defer rows.Close()

	taskRuns := []models.TaskRun{}

	for rows.Next() {
		taskRun := models.TaskRun{}

		var namespace string
		var pipeline string
		var run int64
		var id string
		var taskJSON string
		var created int64
		var started int64
		var ended int64
		var exitCodeRaw sql.NullInt64
		var state string
		var status string
		var statusReasonJSON sql.NullString
		var logsExpired bool
		var logsRemoved bool
		var variablesJSON string

		err = rows.Scan(&namespace, &pipeline, &run, &id, &taskJSON, &created, &started, &ended,
			&exitCodeRaw, &state, &status, &statusReasonJSON, &logsExpired, &logsRemoved, &variablesJSON)
		if err != nil {
			return nil, fmt.Errorf("database error occurred: %v; %w", err, ErrInternal)
		}

		var statusReason *models.TaskRunStatusReason = nil
		if statusReasonJSON.Valid {
			err := json.Unmarshal([]byte(statusReasonJSON.String), statusReason)
			if err != nil {
				return nil, fmt.Errorf("database error occurred; could not decode object; %v", err)
			}
		}

		var exitCode *int64 = nil
		if exitCodeRaw.Valid {
			exitCode = &exitCodeRaw.Int64
		}

		task := models.Task{}
		err = json.Unmarshal([]byte(taskJSON), &task)
		if err != nil {
			return nil, fmt.Errorf("database error occurred; could not decode object; %v", err)
		}

		variables := []models.Variable{}
		err = json.Unmarshal([]byte(variablesJSON), &variables)
		if err != nil {
			return nil, fmt.Errorf("database error occurred; could not decode object; %v", err)
		}

		taskRun.Namespace = namespace
		taskRun.Pipeline = pipeline
		taskRun.Run = run
		taskRun.ID = id
		taskRun.Task = task
		taskRun.Created = created
		taskRun.Started = started
		taskRun.Ended = ended
		taskRun.ExitCode = exitCode
		taskRun.State = models.TaskRunState(state)
		taskRun.Status = models.TaskRunStatus(status)
		taskRun.StatusReason = statusReason
		taskRun.LogsExpired = logsExpired
		taskRun.LogsRemoved = logsRemoved
		taskRun.Variables = variables

		taskRuns = append(taskRuns, taskRun)
	}
	err = rows.Err()
	if err != nil {
		return nil, fmt.Errorf("database error occurred: %v; %w", err, ErrInternal)
	}

	return taskRuns, nil
}

func (db *DB) InsertTaskRun(taskRun *models.TaskRun) error {
	var statusReasonJSON *string = nil
	if taskRun.StatusReason != nil {
		rawJSON, err := json.Marshal(taskRun.StatusReason)
		if err != nil {
			return fmt.Errorf("database error occurred; could not encode object; %v", err)
		}

		statusReasonJSON = ptr(string(rawJSON))
	}

	taskJSON, err := json.Marshal(taskRun.Task)
	if err != nil {
		return fmt.Errorf("database error occurred; could not encode object; %v", err)
	}

	variablesJSON, err := json.Marshal(taskRun.Variables)
	if err != nil {
		return fmt.Errorf("database error occurred; could not encode object; %v", err)
	}

	_, err = qb.Insert("task_runs").Columns("namespace", "pipeline", "run", "id", "created", "started", "ended",
		"exit_code", "logs_expired", "logs_removed", "state", "status", "status_reason", "task", "variables").Values(
		taskRun.Namespace, taskRun.Pipeline, taskRun.Run, taskRun.ID, taskRun.Created, taskRun.Started,
		taskRun.Ended, taskRun.ExitCode, taskRun.LogsExpired, taskRun.LogsRemoved, taskRun.State, taskRun.Status,
		statusReasonJSON, string(taskJSON), string(variablesJSON)).RunWith(db).Exec()
	if err != nil {
		if strings.Contains(err.Error(), "UNIQUE constraint failed") {
			return ErrEntityExists
		}

		return fmt.Errorf("database error occurred: %v; %w", err, ErrInternal)
	}

	return nil
}

func (db *DB) GetTaskRun(namespace, pipeline string, run int64, taskRun string) (models.TaskRun, error) {
	row := qb.Select("namespace", "pipeline", "run", "id", "task", "created", "started", "ended", "exit_code",
		"state", "status", "status_reason", "logs_expired", "logs_removed", "variables").
		From("task_runs").
		Where(qb.Eq{"namespace": namespace, "pipeline": pipeline, "run": run, "id": taskRun}).RunWith(db).QueryRow()

	var namespaceID string
	var pipelineID string
	var runID int64
	var id string
	var taskJSON string
	var created int64
	var started int64
	var ended int64
	var exitCodeRaw sql.NullInt64
	var state string
	var status string
	var statusReasonJSON sql.NullString
	var logsExpired bool
	var logsRemoved bool
	var variablesJSON string

	err := row.Scan(&namespaceID, &pipelineID, &runID, &id, &taskJSON, &created, &started, &ended,
		&exitCodeRaw, &state, &status, &statusReasonJSON, &logsExpired, &logsRemoved, &variablesJSON)
	if err != nil {
		if errors.Is(err, sql.ErrNoRows) {
			return models.TaskRun{}, ErrEntityNotFound
		}

		return models.TaskRun{}, fmt.Errorf("database error occurred: %v; %w", err, ErrInternal)
	}

	var statusReason *models.TaskRunStatusReason = nil
	if statusReasonJSON.Valid {
		err := json.Unmarshal([]byte(statusReasonJSON.String), statusReason)
		if err != nil {
			return models.TaskRun{}, fmt.Errorf("database error occurred; could not decode object; %v", err)
		}
	}

	task := models.Task{}
	err = json.Unmarshal([]byte(taskJSON), &task)
	if err != nil {
		return models.TaskRun{}, fmt.Errorf("database error occurred; could not decode object; %v", err)
	}

	var exitCode *int64 = nil
	if exitCodeRaw.Valid {
		exitCode = &exitCodeRaw.Int64
	}

	variables := []models.Variable{}
	err = json.Unmarshal([]byte(variablesJSON), &variables)
	if err != nil {
		return models.TaskRun{}, fmt.Errorf("database error occurred; could not decode object; %v", err)
	}

	retrievedTaskRun := models.TaskRun{}

	retrievedTaskRun.Namespace = namespace
	retrievedTaskRun.Pipeline = pipeline
	retrievedTaskRun.Run = run
	retrievedTaskRun.ID = id
	retrievedTaskRun.Task = task
	retrievedTaskRun.Created = created
	retrievedTaskRun.Started = started
	retrievedTaskRun.Ended = ended
	retrievedTaskRun.ExitCode = exitCode
	retrievedTaskRun.State = models.TaskRunState(state)
	retrievedTaskRun.Status = models.TaskRunStatus(status)
	retrievedTaskRun.StatusReason = statusReason
	retrievedTaskRun.LogsExpired = logsExpired
	retrievedTaskRun.LogsRemoved = logsRemoved
	retrievedTaskRun.Variables = variables

	return retrievedTaskRun, nil
}

func (db *DB) UpdateTaskRun(taskRun *models.TaskRun, fields UpdatableTaskRunFields) error {
	query := qb.Update("task_runs")

	if fields.Started != nil {
		query = query.Set("started", fields.Started)
	}

	if fields.Ended != nil {
		query = query.Set("ended", fields.Ended)
	}

	if fields.ExitCode != nil {
		query = query.Set("exit_code", fields.ExitCode)
	}

	if fields.State != nil {
		query = query.Set("state", fields.State)
	}

	if fields.Status != nil {
		query = query.Set("status", fields.Status)
	}

	if fields.StatusReason != nil {
		statusReason, err := json.Marshal(fields.StatusReason)
		if err != nil {
			return fmt.Errorf("database error occurred; could not encode object; %v", err)
		}
		query = query.Set("status_reason", statusReason)
	}

	if fields.LogsExpired != nil {
		query = query.Set("logs_expired", fields.LogsExpired)
	}

	if fields.LogsRemoved != nil {
		query = query.Set("logs_removed", fields.LogsRemoved)
	}

	if fields.Variables != nil {
		variables, err := json.Marshal(fields.Variables)
		if err != nil {
			return fmt.Errorf("database error occurred; could not encode object; %v", err)
		}
		query = query.Set("variables", string(variables))
	}

	_, err := query.Where(qb.Eq{
		"namespace": taskRun.Namespace, "pipeline": taskRun.Pipeline, "run": taskRun.Run,
		"id": taskRun.ID,
	}).RunWith(db).Exec()
	if err != nil {
		if errors.Is(err, sql.ErrNoRows) {
			return ErrEntityNotFound
		}

		return fmt.Errorf("database error occurred: %v; %w", err, ErrInternal)
	}

	return nil
}

func (db *DB) DeleteTaskRun(namespace, pipeline string, run int64, id string) error {
	_, err := qb.Delete("task_runs").
		Where(qb.Eq{"namespace": namespace, "pipeline": pipeline, "run": run, "id": id}).RunWith(db).Exec()
	if err != nil {
		if errors.Is(err, sql.ErrNoRows) {
			return nil
		}

		return fmt.Errorf("database error occurred: %v; %w", err, ErrInternal)
	}

	return nil
}
